gapir/client/client.go

```diff
@@ -102,33 +140,152 @@ func (client *Client) Connect(ctx context.Context, device bind.Device, abi *devi
+		grpc.WithInsecure(),
+		grpc.WithBlock(),
+		grpc.WithTimeout(gRPCConnectTimeout),
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(math.MaxInt32)))
 	if err != nil {
 		return nil, log.Err(ctx, err, "Timeout waiting for connection")
 	}
+	rpcClient := replaysrv.NewGapirClient(conn)
 
-	crash.Go(func() { client.heartbeat(ctx, heartbeatInterval, key) })
+	replayerInfo := &replayerInfo{
+		deviceConnectionInfo: *newDeviceConnectionInfo,
+		device:               device,
+		abi:                  abi,
+		conn:                 conn,
+		rpcClient:            rpcClient,
+	}
 
+	crash.Go(func() { client.heartbeat(ctx, replayerInfo) })
 	log.I(ctx, "Heartbeat connection setup done")
 
-	bgConnection, err := client.makeBackgroundConnection(ctx, device, connection)
+	err = replayerInfo.startReplayCommunicationHandler(ctx)
 	if err != nil {
-		return nil, log.Err(ctx, err, "Background connection error")
+		return nil, log.Err(ctx, err, "Error in startReplayCommunicationHandler")
 	}
 
-	client.clientInfos[key] = clientInfo{
-		deviceConnectionInfo: *newDeviceConnectionInfo,
-		connection:           connection,
-		device:               device,
-		arch:                 abi.Architecture,
-		abi:                  abi,
-		bgConnection:         bgConnection}
+	client.replayers[key] = replayerInfo
 	return &key, nil
 }
 
-func (client *Client) makeBackgroundConnection(ctx context.Context, device bind.Device, conn gapir.Connection) (*backgroundConnection, error) {
-	bgc := &backgroundConnection{conn: conn, OS: device.Instance().GetConfiguration().GetOS()}
+// removeConnection closes and removes a GAPIR instance.
+func (client *Client) removeConnection(ctx context.Context, key ReplayerKey) {
+	client.mutex.Lock()
+	defer client.mutex.Unlock()
+
+	if replayer, ok := client.replayers[key]; ok {
+		replayer.closeConnection(ctx)
+		delete(client.replayers, key)
+	}
+}
+
+// reconnect removes the replayer before re-creating it.
+func (client *Client) reconnect(ctx context.Context, replayer *replayerInfo) {
+	key := ReplayerKey{device: replayer.device, arch: replayer.abi.GetArchitecture()}
+	client.removeConnection(ctx, key)
+	client.Connect(ctx, replayer.device, replayer.abi)
+}
+
+// heartbeat regularly sends a ping to a replayer, and restarts it when it fails to reply.
+func (client *Client) heartbeat(ctx context.Context, replayer *replayerInfo) {
+	for {
+		select {
+		case <-task.ShouldStop(ctx):
+			return
+		case <-time.After(heartbeatInterval):
+			err := replayer.ping(ctx)
+			if err != nil {
+				log.E(ctx, "Error sending keep-alive ping. Error: %v", err)
+				client.reconnect(ctx, replayer)
+				return
+			}
+		}
+	}
+}
+
+// getActiveReplayer returns the replayer identified by key only if this
+// replayer has an active connection to a GAPIR instance.
+func (client *Client) getActiveReplayer(ctx context.Context, key *ReplayerKey) (*replayerInfo, error) {
+	replayer, found := client.replayers[*key]
```

Comment:
This function should be protected by mutex.

If the heartbeat calls reconnect from a channel(I don't know the Go's channel model well but I suspect that it can be from another thread), when this method is in process, we may crash in for example in line 217 due to replayer being deleted meanwhile.

Reply:
I guess the purpose of this PR is not to fix existing issues, just to move code around and remove complexity. I see however a benefit of following up quickly on this feedback to actually fix the issues we're finding here.